core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -287,6 +287,17 @@ private[spark] class TaskSchedulerImpl(
     }
   }
 
+  override def completeTasks(partitionId: Int, stageId: Int, taskInfo: TaskInfo): Unit = {
+    taskSetsByStageIdAndAttempt.getOrElse(stageId, Map()).values.foreach { tsm =>
+      tsm.partitionToIndex.get(partitionId) match {
+        case Some(index) =>
+          tsm.markPartitionCompleted(index, taskInfo)
```

Comment:
partition and index seem mixed up here.  They are usually the same on the first stage attempt, but are almost always different later on, so they need to be treated with care.  `tsm.markPartitionCompleted` expects a partition currently.  I think you should just skip doing the `tsm.partitionToIndex.get(partitionId)` here.

Reply:
Yikes... this was my bad..... have updated the code. Thank you for catching that.